docs/development/design/watch-api.md

```diff
@@ -0,0 +1,191 @@
+# Nexodus Device Watch API Implementation Notes
+
+## The Problem
+
+In order to connect or disconnect devices from a Nexodus managed organization, the list of all desired devices in the organization need replicated to all devices in the organization.  The `ListDevicesInOrganization` operation provides this list and can be accessed via the following REST call:
+
+`GET /api/organizations/{organization_id}/devices`
+
+The `nexd` agent would need to periodically poll this endpoint to get the current desired state from the API server and reconcile the device configuration to add or remove Wireguard peers endpoints as nessary.
+
+As the number of devices (N) in the organization increases, the the number of periodic API requests to `ListDevicesInOrganization` increases linearly.  The size of the reponse list also increases linearly.  Put those two together and you get the reponse bandwith to the API server increasing exponentially (N requests/polling interval x N device elements in the response).  This will clearly lead to scaling challenges with large organizations.
+
+This also has the problem that desired state changes recorded in the apiserver will not take effect immediatly since the `nexd` agents are likely in the middle of their poll interval, and wont poll the API server for a little while.
+
+## The Solution: The Watch style API
+
+Inspired by the Kubernetes List/Watch style APIs, the Nexodus `ListDevicesInOrganization` operation can be called with the `watch=true` query parameter.  Furthemore devices have been given a `revision` field that can be used to efficently make change tracking of devices possible.
+
+When you call the `ListDevicesInOrganization` operation with the `watch=true` query parameter, the API server responds with a stream of changes. These changes itemize the outcome of operations (such as **change**, **delete**, and **update**) that occurred after the `revision` you specified as a query parameter to the request. The overall **watch** mechanism allows a client to fetch the current state and then subscribe to subsequent changes, without missing any events.
+
+If a client **watch** is disconnected then that client can start a new **watch** from the last returned `revision`; the client could also perform a fresh watch without specifying a revision and begin again.
+
+This has the effect of having better bandwith scaling characteristics since only change events are sent to N devices when the desired state changes.  If the devices are not actively joining and leaving the organization, this can be verly little bandwidth.  And even if they are joining and leaving at a constant rate, the bandwidth scales with N (not N x N).  In addition to the bandwidth savings, peer devices are pushed change events as soon as possible for processing which means they can converge on the desired state quicker than waiting for the next polling interval.
```

Comment:
```suggestion
This has the effect of having better bandwidth scaling characteristics since only change events are sent to N devices when the desired state changes.  If the devices are not actively joining and leaving the organization, this can be very little bandwidth.  And even if they are joining and leaving at a constant rate, the bandwidth scales with N (not N x N).  In addition to the bandwidth savings, peer devices are pushed change events as soon as possible for processing which means they can converge on the desired state quicker than waiting for the next polling interval.
```